File: C++/rider.cpp

#include "rider.h"

Rider::Rider(int id, string riderName)
    : riderID(id), name(riderName) {}

void Rider::requestRide(shared_ptr<Ride> ride) {
    requestedRides.push_back(ride);
    cout << "  Rider " << name << " requested Ride #"
         << ride->getRideID() << " (" << ride->getRideType() << ")" << endl;
}

void Rider::viewRides() {
    cout << endl;

    cout << "Rider history:" << endl;

    cout << "  Rider ID     : " << riderID << endl;
    cout << "  Name         : " << name << endl;
    cout << "  Total Rides  : " << requestedRides.size() << endl;

    if (!requestedRides.empty()) {
        double totalSpent = 0.0;
        cout << endl << " Requested Rides" << endl;
        for (auto& ride : requestedRides) {
            ride->rideDetails();
            totalSpent += ride->fare();
        }
        cout << "  Total Spent    : $" << fixed << setprecision(2)
             << totalSpent << endl;
    }
    cout << endl;
}

int Rider::getRiderID() const { return riderID; }
string Rider::getName() const { return name; }

File: Go/rider.go
package main

import "fmt"

// Rider

type Rider struct {
	riderID        int
	name           string
	requestedRides []Ride 
}

func NewRider(id int, name string) *Rider {
	return &Rider{
		riderID:        id,
		name:           name,
		requestedRides: []Ride{},
	}
}

func (r *Rider) RequestRide(ride Ride) {
	r.requestedRides = append(r.requestedRides, ride)
	fmt.Printf("  [+] Rider %s requested Ride #%d (%s)\n",
		r.name, ride.GetRideID(), ride.RideType())
}

func (r *Rider) ViewRides() {
	fmt.Println()

	fmt.Println("Rider history")

	fmt.Printf("  Rider ID     : %d\n", r.riderID)
	fmt.Printf("  Name         : %s\n", r.name)
	fmt.Printf("  Total Rides  : %d\n", len(r.requestedRides))

	if len(r.requestedRides) > 0 {
		totalSpent := 0.0
		fmt.Println()
		fmt.Println(" Requested Rides ")
		for _, ride := range r.requestedRides {
			ride.RideDetails()
			totalSpent += ride.Fare()
		}
		fmt.Printf("  Total Spent    : $%.2f\n", totalSpent)
	}

}
